optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,350 @@
+            "to `False` in the constructor of `QMCSampler`, "
+            "if this independent sampling is intended behavior.".format(
+                param_name, trial.number, self._independent_sampler.__class__.__name__
+            )
+        )
+
+    @staticmethod
+    def _log_asyncronous_seeding() -> None:
+        _logger.warning(
+            "No seed is provided for `QMCSampler` and the seed is set randomly. "
+            "If you are running multiple `QMCSampler`s in parallel and/or distributed "
+            " environment, the same seed must be used in all samplers to ensure that resulting "
+            "samples are taken from the same QMC sequence. "
+            "You can suppress this warning by setting `warn_asyncronous_seeding` "
+            "to `False` in the constructor of `QMCSampler`, "
+            "if this random seeding is intended behavior."
+        )
+
+    def sample_independent(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        param_name: str,
+        param_distribution: BaseDistribution,
+    ) -> Any:
+
+        if self._initial_search_space is not None:
+            if self._warn_independent_sampling:
+                self._log_independent_sampling(trial, param_name)
+
+        return self._independent_sampler.sample_independent(
+            study, trial, param_name, param_distribution
+        )
+
+    def sample_relative(
+        self, study: Study, trial: FrozenTrial, search_space: Dict[str, BaseDistribution]
+    ) -> Dict[str, Any]:
+
+        if search_space == {}:
+            return {}
+
+        sample = self._sample_qmc(study, trial, search_space)
+        trans = _SearchSpaceTransform(search_space)
+        sample = scipy.stats.qmc.scale(sample, trans.bounds[:, 0], trans.bounds[:, 1])
+        sample = trans.untransform(sample[0, :])
+        return sample
+
+    def _sample_qmc(
+        self, study: Study, trial: FrozenTrial, search_space: Dict[str, BaseDistribution]
+    ) -> numpy.ndarray:
+
+        # Lazy import because the `scipy.stats.qmc` is slow to import.
+        import scipy.stats.qmc
+
+        sample_id = self._find_sample_id(study, trial, search_space)
+        d = len(search_space)
+
+        # Use cached `qmc_engine` or construct a new one.
+        if self._is_engine_cached(d, sample_id):
+            qmc_engine = self._cached_qmc_engine
+        else:
+            if self._qmc_type == "sobol":
+                qmc_engine = scipy.stats.qmc.Sobol(d, seed=self._seed, scramble=self._scramble)
+            elif self._qmc_type == "halton":
+                qmc_engine = scipy.stats.qmc.Halton(d, seed=self._seed, scramble=self._scramble)
+            else:
+                message = (
+                    f"The `qmc_type`, {self._qmc_type}, is not a valid. "
+                    'It must be one of "sobol" and "halton".'
+                )
+                raise ValueError(message)
+
+        assert isinstance(qmc_engine, scipy.stats.qmc.QMCEngine)
```

Comment:
Is this intended? This can break as running python with `-O` would remove `assert` statements.

Reply:
I added this line to get rid of the error of `mypy` (static type checker). The code will behave the same without this assertion.

Reply:
ok, you could also add it when defining the variables in that case. `qmc_engine: scipy.stats.qmc.QMCEngine = ...`

Reply:
Thank you, it worked OK without the assertion if I typed the `qmc_engine` as you suggested!